platform/linux-generic/arch/aarch64/odp_crypto_armv8.c

```diff
@@ -137,6 +154,280 @@ null_crypto_routine(odp_packet_t pkt ODP_UNUSED,
 	return ODP_CRYPTO_ALG_ERR_NONE;
 }
 
+static
+odp_crypto_alg_err_t aes_gcm_encrypt(odp_packet_t pkt,
+				     const odp_crypto_packet_op_param_t *param,
+				     odp_crypto_generic_session_t *session)
+{
+	armv8_cipher_state_t cs = {
+		.counter = {
+			.d = {0, 0}
+		}
+	};
+	armv8_cipher_constants_t cc = {
+		.mode = 0
+	};
+	uint8_t *iv_ptr;
+	uint64_t nonce_length, nonce_byte_length;
+	uint64_t plaintext_length, plaintext_byte_length;
+	uint64_t aad_length, aad_byte_length;
+	int ret = 0;
+
+	/* Clear the allocated buffers */
+	memset(global->input, 0, _ARM_CRYPTO_MAX_DATA_LENGTH);
+	memset(global->nonce, 0, _ARM_CRYPTO_MAX_IV_LENGTH);
+	memset(global->aad, 0, _ARM_CRYPTO_MAX_AAD_LENGTH);
+	memset(global->digest, 0, _ARM_CRYPTO_MAX_DIGEST_LENGTH);
+	memset(global->output, 0, (_ARM_CRYPTO_MAX_DATA_LENGTH +
```

Comment:
This does not work. Since ODP and ODP apps are multithreaded and the crypto operations can be called simultaneously by multiple threads, you cannot modify local data just like that. Temporary scratch buffers like that should be thread local (or protected by locks, but that would really slow things down).

Reply:
Currently these buffers are declared as a part of the global structure, odp_crypto_global_t. I can move these buffers to the thread local structure, crypto_local_t and then initialize them. Or I can only initialize them once as a part of _odp_crypto_init_global and remove these lines from the function. What would be preferable?

Reply:
> Or I can only initialize them once as a part of _odp_crypto_init_global and remove these lines from the function. 

That would not fix the problem. The problem is that you cannot store temporary per-thread data in a global structure where different threads would stomp over each other in writing and reading it. IOW, the later R/W accesses to the global struct are also wrong. The initialization itself actually seems unnecessary at a quick look.

Reply:
Okay, understood. I will move them to the thread local struct and get rid of the initialization as well.